From c9fe630e728ef1e2cdac72b3b3f0ec954d76c7e2 Mon Sep 17 00:00:00 2001
From: Wenzong Fan <wenzong.fan@windriver.com>
Date: Tue, 2 Jan 2018 07:46:41 +0000
Subject: [PATCH] Add userdom_transition_login_userdomain() interface

Port 3 patches from https://github.com/fedora-selinux/selinux-policy:
  70e661b3 Add userdom_transition_login_userdomain() interface
  0d49b414 Add login_userdomain attribute for users which
           can log in using terminal
  51af8382 Call userdom_transition_login_userdomain() \
           instead of userdom_transition() in init.te \
           related to pam_selinux.so+systemd-users.

Fix avc denial:
  avc: denied { transition } for pid=647 comm="(systemd)" \
  path="/lib/systemd/systemd" dev="sda1" ino=1325579 \
  scontext=system_u:system_r:init_t \
  tcontext=root:sysadm_r:sysadm_t tclass=process

Upstream-Status: Pending

Signed-off-by: Wenzong Fan <wenzong.fan@windriver.com>
---
 policy/modules/system/init.te       |  2 ++
 policy/modules/system/userdomain.if | 21 +++++++++++++++++++++
 policy/modules/system/userdomain.te |  1 +
 3 files changed, 24 insertions(+)

diff --git a/policy/modules/system/init.te b/policy/modules/system/init.te
index 43b1532..2005eee 100644
--- a/policy/modules/system/init.te
+++ b/policy/modules/system/init.te
@@ -169,6 +169,8 @@ domain_signull_all_domains(init_t)
 domain_sigstop_all_domains(init_t)
 domain_sigchld_all_domains(init_t)
 
+userdom_transition_login_userdomain(init_t)
+
 files_read_etc_files(init_t)
 files_rw_generic_pids(init_t)
 files_manage_etc_runtime_files(init_t)
diff --git a/policy/modules/system/userdomain.if b/policy/modules/system/userdomain.if
index ad64591..fce30a1 100644
--- a/policy/modules/system/userdomain.if
+++ b/policy/modules/system/userdomain.if
@@ -237,6 +237,24 @@ template(`userdom_user_content_access_template',`
 	')
 ')
 
+########################################
+## <summary>
+##     Allow caller to transition to login userdomain.
+## </summary>
+## <param name="domain">
+##     <summary>
+##     Domain allowed access.
+##     </summary>
+## </param>
+#
+interface(`userdom_transition_login_userdomain',`
+	gen_require(`
+		attribute login_userdomain;
+	')
+
+	allow $1 login_userdomain:process transition;
+')
+
 #######################################
 ## <summary>
 ##	Allow a home directory for which the
@@ -896,10 +914,13 @@ template(`userdom_common_user_template',`
 template(`userdom_login_user_template', `
 	gen_require(`
 		class context contains;
+		attribute login_userdomain;
 	')
 
 	userdom_base_user_template($1)
 
+	typeattribute $1_t login_userdomain;
+
 	userdom_manage_home_role($1_r, $1_t)
 
 	userdom_manage_tmp_role($1_r, $1_t)
diff --git a/policy/modules/system/userdomain.te b/policy/modules/system/userdomain.te
index 0ec2461..5c14dcc 100644
--- a/policy/modules/system/userdomain.te
+++ b/policy/modules/system/userdomain.te
@@ -66,6 +66,7 @@ gen_tunable(user_write_removable, false)
 gen_tunable(user_ttyfile_stat, false)
 
 attribute admindomain;
+attribute login_userdomain;
 
 # all user domains
 attribute userdomain;
-- 
2.7.4

